ISLANDORA-2172: Take advantage of content negociation capabilities

d-r-p opened the pull request:
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2172

# What does this Pull Request do?

Currently, [islandora_doi](https://github.com/Islandora/islandora_scholar/tree/7.x/modules/doi) uses [Crossref](https://www.crossref.org)'s openurl to retrieve metadata. This requires users to set up an account with Crossref.

Instead, islandora_doi can take advantage of [content negociation](https://www.crosscite.org/docs.html) and retrieve the metadata directly by querying https://doi.org. Not only does this method work without prior account creation, it also has the benefit of working with other metadata providers (e.g. [DataCite](https://datacite.org), [mEDRA](https://www.medra.org), etc...). This makes islandora_doi fit for future extensions.

This pull request adds this possibility, along with improving error messages.

# What's new?

* Introduce a new mode for retrieving metadata featuring:
   * Use content negociation rather than openurl
   * No prior account creation required
   * Prepare for extending to metadata providers other than Crossref
      (e.g. DataCite, mEDRA)
* Streamline error handling:
   * Handle errors exclusively in modules/doi/include/utilities.inc in order to
      obtain same messages regardless of ingest method (populator/importer)
   * Especially in the new non-legacy mode, failures to ingest should now give
      more useful messages (wrong type, wrong metadata provider, etc...)

The changes should be pretty non-intrusive. In particular, all existing
functions have been preserved (no API change for those who created their own
extensions). Also, users who already have their account set up with Crossref
will default to keep querying in legacy mode. Only the improvements in error
handling should be visible. New users will default to querying in the new mode.
The querying behaviour can be changed at all times on the configuration page
/admin/islandora/solution_pack_config/scholar/islandora_doi.


# How should this be tested?

As always, never test in a production environment!

Steps:
* Test importer & populator on different DOIs (including DOIs of metadata providers others than Crossref, as well as other document types such as books)
* Merge this PR and re-test. Everything but the error messages should be the same (the latter being a bit more useful)
* Go to /admin/islandora/solution_pack_config/scholar/islandora_doi and de-select legacy mode
* Re-test and see better error messages
* Delete the variable 'islandora_doi_openurl_pid' (or set the PID to the default "user@example.com") and re-test: in non-legacy mode the metadata is still retrievable

# Additional Notes:

* Does this change require documentation to be updated? **no**
* Does this change add any new dependencies? **no**
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? **no**
* Could this change impact execution of existing code? **no**

# Interested parties
@bryjbrown @DonRichards @Islandora/7-x-1-x-committers

d-r-p committed:
ISLANDORA-2172: Take advantage of content negociation capabilities of https://doi.org

- Introduce a new mode for retrieving metadata featuring:
  - Use content negociation rather than openurl
  - No prior account creation required
  - Prepare for extending to metadata providers other than Crossref
    (e.g. DataCite, mEDRA)
- Streamline error handling:
  - Handle errors exclusively in modules/doi/include/utilities.inc in order to
    obtain same messages regardless of ingest method (populator/importer)
  - Especially in the new non-legacy mode, failures to ingest should now give
    more useful messages (wrong type, wrong metadata provider, etc...)

The changes should be pretty non-intrusive. In particular, all existing
functions have been preserved (no API change for those who created their own
extensions). Also, users who already have their account set up with Crossref
will default to keep querying in legacy mode. Only the improvements in error
handling should be visible. New users will default to querying in the new mode.
The querying behaviour can be changed at all times on the configuration page
/admin/islandora/solution_pack_config/scholar/islandora_doi.

d-r-p commented:
Regarding Travis CI failures: sorry for the hiccups, I will sanitise the code later today.

d-r-p committed:
Sanitise code

- Coding standards
- Safe usage of watchdog(), trigger_error(), t(), drupal_set_message()

adam-vessey commented on modules/doi/includes/utilities.inc:
```diff
@@ -5,6 +5,48 @@
+global $_islandora_doi_content_validators;
+$_islandora_doi_content_validators = array(
+  'application/vnd.crossref.unixref+xml' => 'islandora_doi_crossref_validator',
+);
```
These two global variables would probably be better off being built up from hooks... Really, could see it as one hook... a `hook_islandora_doi_content_type_info()` or similar... This module's implementation would be something like:
```php
function islandora_doi_islandora_doi_content_type_info() {
  return array(
    'application/vnd.crossref.unixref+xml' => array(
      'translator' => 'islandora_doi_crossref_translator',
      'validator' => 'islandora_doi_crossref_validator',
    ),
  );
}
```

Given some of the [inline documentation](https://github.com/Islandora/islandora_scholar/pull/280/files#diff-52bb712f031b7c5151fc17f7508f7213R26), it looks like it may also be necessary to add in a "weight" parameter, and perform the ordering after the hook is invoked... may be useful to roll a utility function to perform the hook invocation and ordering, and probably a `drupal_alter()` invocation on the result for good measure.

It may be desirable to add some other properties, such as `file` to allow other files to be specified, so not all callbacks need to exist in `.module` files... if we do so, it may be convenient to roll some other utility functions.

... Would still require the hook and callback signatures to be documented in the `.api.php`, and the hook to be invoked where appropriate.

```php
/**
 * Gather info for content negotiation.
 *
 * @return array
 *   An associative array mapping MIME-types to associative arrays containing:
 *    - translator: A callable implementing callback_islandora_doi_translator().
 *    - validator: A callable implementing callback_islandora_doi_validator().
 */
function hook_islandora_doi_content_type_info() {
  
}

/**
 * Translate response to MODS, contained in a DOMDocument.
 *
 * @param DOMDocument $doc
 *   The response to be transformed.
 *
 * @return DOMDocument|bool
 *   The translated document as a DOMDocument; otherwise, boolean FALSE on failure.
 */
function callback_islandora_doi_translator(DOMDocument $doc) {
  // For an example, see islandora_doi_crossref_translator().
}

/**
 * Validate response.
 *
 * @param string $id
 *   The DOI the response is supposed to represent.
 * @param string $data
 *   The response to validate.
 *
 * @return bool
 *   TRUE if the response appears to be valid; otherwise, FALSE.
 */
function callback_islandora_doi_validator($id, $data) {
  // For an example, see islandora_doi_crossref_validator().
}
```

Basing the callback signatures based the this PR; though... I feel it may be desirable to change the translator to accept the raw response, such that non-XML response formats could hypothetically be handled.

adam-vessey commented on modules/doi/includes/utilities.inc:
```diff
@@ -555,30 +586,202 @@ function islandora_doi_transform_to_mods($doi_xml) {
     return FALSE;
   }
+
+  return array($content_type => $response->data);
```
Instead of returning an associative array, it may be simpler to return a flat array...

```php
return array($content_type, $response->data);
```
 
Then [below](https://github.com/Islandora/islandora_scholar/pull/280/files#diff-52bb712f031b7c5151fc17f7508f7213R780), when the two values need to be read, the `list()` operator can be used:
```php
$values = islandora_doi_get_source($id);
if (!values) {
  return FALSE;
}
list($content_type, $source_string) = $values;
```

adam-vessey commented on modules/doi/modules/doi_importer/includes/importer.inc:
```diff
@@ -117,6 +118,13 @@ class DOIImportObject extends IslandoraImportObject {
+   *
+   * @var string
+   */
+  protected $sourceContentType;
```
The introduction of this new member will effectively break any existing items in the batch queue...

... Could address either in code (populate on unserialization or something of the like) or via documentation.

DiegoPino replied:
+1 to documenting.

adam-vessey replied:
[Appears to be addressed](https://github.com/Islandora/islandora_scholar/pull/280/files#diff-2d5c3db3bf912797bde525b055da1cfbR162).

adam-vessey commented on modules/doi/includes/admin.form.inc:
```diff
@@ -18,17 +18,39 @@
+  $form['islandora_doi_url_config'] += array(
+    'islandora_doi_doiurl' => array(
+      '#type' => 'textfield',
+      '#title' => t('DOI URL supporting content negociation'),
```
`s/negociation/negotiation/`

adam-vessey commented on modules/doi/includes/utilities.inc:
```diff
@@ -5,6 +5,48 @@
+ * functions are supposed to translate the returned XML data into MODS.
+ *
+ * Note: At the time of writing, @link https//doi.org doi.org @endlink was able
+ * to perform the content negociation as explained on
```
`s/negociation/negotiation/`

adam-vessey commented on modules/doi/modules/doi_importer/README.md:
```diff
@@ -19,7 +19,7 @@ Install as usual, see [this](https://drupal.org/documentation/install/modules-th
 ## Configuration
 
-Set the value for `OpenURL PID` at `admin/islandora/scholar/doi_importer` as CrossRef does not allow guest searching.
+Set the value for `OpenURL PID` at `admin/islandora/scholar/doi_importer` as CrossRef does not allow guest searching, or just use the new content negociation based metadata retrieval.
```
`s/negociation/negotiation/`

adam-vessey replied:
It seems like this could be made more to the point:
```
By default, no configuration is required.

If you wish to use the legacy CrossRef OpenURL endpoint, you must enable it 
and enter your `OpenURL PID` at `admin/islandora/scholar/doi_importer`.

It is possible to change the URL against which content negotiation requests are made by
changing the `DOI URL supporting content negotiation` URL at
`admin/islandora/scholar/doi_importer`.
```

adam-vessey commented on modules/doi/includes/admin.form.inc:
```diff
@@ -18,17 +18,39 @@
+    'islandora_doi_use_legacy_mode' => array(
+      '#type' => 'checkbox',
+      '#title' => t('Use legacy retrieval mode'),
+      '#description' => t("Keep fetching metadata directly from CrossREF's OpenURL (account required!). If unchecked, use content negociation as explained by https://www.crosscite.org."),
```
`s/negociation/negotiation/`

adam-vessey commented on modules/doi/modules/doi_importer/includes/importer.inc:
```diff
@@ -25,16 +25,17 @@ class DOIImporter extends IslandoraBatchImporter {
-    if ($openurl_pid == 'user@example.com') {
-      drupal_set_message(t('OpenURL PID email address has not been configured, please enter a valid email address in the <a href="@doi-admin-page">Scholar DOI configuration form</a>.', array('@doi-admin-page' => '/admin/islandora/solution_pack_config/scholar/islandora_doi')), 'error');
+    $openurl_pid = variable_get('islandora_doi_openurl_pid', ISLANDORA_DOI_OPENURL_PID_DEFAULT);
+    $legacy_mode = variable_get('islandora_doi_use_legacy_mode', $openurl_pid !== ISLANDORA_DOI_OPENURL_PID_DEFAULT);
```
There's a number of places where this same test is performed, for getting the "legacy_mode" status, with a dynamic default... probably a good idea to pull it out into a utility function, and call the function where the test needs to be made.

adam-vessey commented on modules/doi/islandora_doi.module:
```diff
@@ -5,6 +5,9 @@
 
+define('ISLANDORA_DOI_OPENURL_DEFAULT', 'http://www.crossref.org/openurl');
+define('ISLANDORA_DOI_OPENURL_PID_DEFAULT', 'user@example.com');
+define('ISLANDORA_DOI_DOIURL_DEFAULT', 'https://doi.org');
```
Very minor point of discussion, but: [Drupal 8 recommends using `const` instead of `define()` "for performance reasons"](https://www.drupal.org/docs/develop/standards/coding-standards#naming)... The likely reason this same recommendation is not made for Drupal 7 is because [Drupal 7 supports PHP versions prior to 5.3](https://www.drupal.org/docs/7/system-requirements/drupal-7-php-requirements), while [it was with 5.3 that the `const` keyword was introduced](http://php.net/const)... Given that [this module is dependent on `islandora`](https://github.com/d-r-p/islandora_scholar/blob/986f6197d3a550335fb1bf365197096ec71cd45e/modules/doi/islandora_doi.info#L4), which in turn [requires 5.3 at minimum](https://github.com/Islandora/islandora/blob/124b39959f52153ff9dd952bace0fa4e04944c56/islandora.info#L30)... we should be good to proceed with `const`, if we wanted to, though `defined()` is fine.

d-r-p committed:
Implement changes requested by @adam-vessey.

- Fix typos
- Update documentation
- Replace global variables by hook
- Create `.api.php` file
- Have translator functions handle raw data
- Create utility function to retrieve mode
- Change return value of `islandora_doi_get_source()`
- Try to avoid breaking existing batch members
- Change define statements into const statements

Note: Since translator functions now handle raw data
(i.e. strings), we are forced to deprecate the function
`islandora_doi_transform_to_mods()`. Code should,
therefore, be updated to use the new function
`islandora_doi_translate()` instead.

bryjbrown commented:
To be totally honest, I never use the DOI importer in Scholar, so I wouldn't make for a good tester. @DonRichards @dmoses @rosiel @bondjimbond Do any of you import via DOIs and/or would be willing to test this?

adam-vessey commented:
This looks good to me, put things through their paces, with and without the legacy flow:
* forced the pre-existing batch set behaviour (from before to with this code)
* various journal articles imported successfully
* book chapter threw an error, as expected

However... are you covered by a [Contributor license agreeement](https://github.com/Islandora/islandora/blob/7.x/CONTRIBUTING.md#contribute-code), @d-r-p? Not seeing your handle, nor Lib4RI in [the list](https://github.com/Islandora/islandora/wiki/Contributor-License-Agreements).

DiegoPino commented:
He is covered by an iCLA

El El lun, 26 de mar. de 2018 a las 07:27, Adam <notifications@github.com>
escribió:

> This looks good to me, put things through their paces, with and without
> the legacy flow:
>
>    - forced the pre-existing batch set behaviour (from before to with
>    this code)
>    - various journal articles imported successfully
>    - book chapter threw an error, as expected
>
> However... are you covered by a Contributor license agreeement
> <https://github.com/Islandora/islandora/blob/7.x/CONTRIBUTING.md#contribute-code>,
> @d-r-p <https://github.com/d-r-p>? Not seeing your handle, nor Lib4RI in the
> list
> <https://github.com/Islandora/islandora/wiki/Contributor-License-Agreements>
> .
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora/islandora_scholar/pull/280#issuecomment-376185152>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn854fAHuCNq6-yhfO9fSkeYehnQPM_ks5tiPrrgaJpZM4SkNDv>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

adam-vessey commented:
Ah, searching about, [I see it now](https://github.com/Islandora/islandora_xml_forms/pull/247#issuecomment-371605859)... just... the list hasn't been updated?

DiegoPino commented:
Ok, 22 hours! for merge. Thanks @adam-vessey and @d-r-p for coding this

DonRichards commented:
Thanks @adam-vessey
